Malmo/src/MissionSpec.cpp

```diff
@@ -18,656 +18,625 @@
     {
-        AgentHandlers::ColourMapProducer_optional& cps = this->mission->AgentSection().front().AgentHandlers().ColourMapProducer();
-        cps.set(ColourMapProducer(width, height));
+        mission.put("Mission.AgentSection.AgentHandlers.ColourMapProducer.Width", width);
+        mission.put("Mission.AgentSection.AgentHandlers.ColourMapProducer.Height", height);
     }
 
     void MissionSpec::request32bppDepth(int width, int height)
     {
-        AgentHandlers::DepthProducer_optional& dps = this->mission->AgentSection().front().AgentHandlers().DepthProducer();
-        dps.set(DepthProducer(width, height));
+        mission.put("Mission.AgentSection.AgentHandlers.DepthProducer.Width", width);
+        mission.put("Mission.AgentSection.AgentHandlers.DepthProducer.Height", height);
     }
 
     void MissionSpec::requestVideoWithDepth(int width, int height)
     {
-        AgentHandlers::VideoProducer_optional& vps = this->mission->AgentSection().front().AgentHandlers().VideoProducer();
-        VideoProducer vp(width, height);
-        vp.want_depth(true);
-        vps.set(vp);
+        requestVideo(width, height);
+        mission.put("Mission.AgentSection.AgentHandlers.VideoProducer.<xmlattr>.want_depth", true);
     }
     
     void MissionSpec::setViewpoint(int viewpoint)
     {
-        AgentHandlers::VideoProducer_optional& vps = this->mission->AgentSection().front().AgentHandlers().VideoProducer();
-        if( vps.present() ) {
-            vps->viewpoint(viewpoint);
+        const auto& v = mission.get_child_optional("Mission.AgentSection.AgentHandlers.VideoProducer");
+        if (v) {
+            mission.put("Mission.AgentSection.AgentHandlers.VideoProducer.<xmlattr>.viewpoint", viewpoint);
         }
-        // else silently do nothing since no video requested
     }
 
     void MissionSpec::rewardForReachingPosition(float x, float y, float z, float amount, float tolerance)
     {
-        AgentHandlers::RewardForReachingPosition_optional& rrp = this->mission->AgentSection().front().AgentHandlers().RewardForReachingPosition();
-        if (!rrp.present())
-        {
-            RewardForReachingPosition reward;
-            reward.Marker().push_back(PointWithReward(x, y, z, amount, tolerance));
-            rrp.set(reward);
-        }
-        else
-        {
-            rrp->Marker().push_back(PointWithReward(x, y, z, amount, tolerance));
-        }
+        mission.put("Mission.AgentSection.AgentHandlers.RewardForReachingPosition.Marker.<xmlattr>.x", x);
+        mission.put("Mission.AgentSection.AgentHandlers.RewardForReachingPosition.Marker.<xmlattr>.y", y);
+        mission.put("Mission.AgentSection.AgentHandlers.RewardForReachingPosition.Marker.<xmlattr>.z", z);
+        mission.put("Mission.AgentSection.AgentHandlers.RewardForReachingPosition.Marker.<xmlattr>.reward", amount);
+        mission.put("Mission.AgentSection.AgentHandlers.RewardForReachingPosition.Marker.<xmlattr>.tolerance", tolerance);
     }
```

Comment:
Again, RewardForReachingPosition has an _array_ of points - this method should append to the array.